libraries/node-core-library/src/FileSystem.ts

```diff
@@ -217,17 +283,49 @@ export class FileSystem {
    * Changes the permissions (i.e. file mode bits) for a filesystem object.
    * Behind the scenes it uses `fs.chmodSync()`.
    * @param path - The absolute or relative path to the object that should be updated.
+   * @param modeBits - UNIX-style file mode bits (e.g. 777 or 666 etc)
+   */
+  public static changePosixModeBits(path: string, mode: PosixModeBits): void {
+    fs.chmodSync(path, mode);
+  }
+
+  /**
+   * Retrieves the permissions (i.e. file mode bits) for a filesystem object.
+   * Behind the scenes it uses `fs.chmodSync()`.
+   * @param path - The absolute or relative path to the object that should be updated.
    * @param mode - UNIX-style file mode bits (e.g. 777 or 666 etc)
    */
-  public static changePermissionBits(path: string, mode: IFileModeBits): void {
-    // tslint:disable-next-line:no-bitwise
-    const modeAsOctal: number = (mode.Owner << 6) + (mode.Group << 3) + (mode.Other);
-    fsx.chmodSync(path, modeAsOctal);
+  public static getPosixModeBits(path: string): PosixModeBits {
+    return FileSystem.getStatistics(path).mode;
+  }
+
+  /**
+   * Returns a 10-character string representation of a PosixModeBits value similar to what
+   * would be displayed by the POSIX command "ls -l".
+   * @remarks
+   * For example, `PosixModeBits.AllRead | PosixModeBits.AllWrite` would be formatted as "-rw-rw-rw-".
+   */
+  public static formatPosixModeBits(mode: PosixModeBits): string {
+    let result: string = '-';  // (later we may add support for additional states such as S_IFDIR or S_ISUID)
+
+    result += (mode & PosixModeBits.UserRead) ? 'r' : '-';
+    result += (mode & PosixModeBits.UserWrite) ? 'w' : '-';
+    result += (mode & PosixModeBits.UserExecute) ? 'x' : '-';
+
+    result += (mode & PosixModeBits.GroupRead) ? 'r' : '-';
+    result += (mode & PosixModeBits.GroupWrite) ? 'w' : '-';
+    result += (mode & PosixModeBits.GroupExecute) ? 'x' : '-';
+
+    result += (mode & PosixModeBits.OthersRead) ? 'r' : '-';
+    result += (mode & PosixModeBits.OthersWrite) ? 'w' : '-';
+    result += (mode & PosixModeBits.OthersExecute) ? 'x' : '-';
```

Comment:
I think this operation is slower than setting up an array and using .join or reduce...

Reply:
The strings are tiny

---
In reply to: [205244478](https://github.com/Microsoft/web-build-tools/pull/754#discussion_r205244478) [](ancestors = 205244478)